src/data/finset/basic.lean

```diff
@@ -1504,6 +1504,231 @@ theorem forall_mem_insert [d : decidable_eq α]
+begin
+  intros A1 A2,
+  intro x,
+  apply le_trans,
+  apply A1,
+  apply A2,
+end
+
+
+lemma classical.some_func {α β:Type*} {P:α → β → Prop}:
+    (∀ a:α, ∃ b:β,  P a b) → (∃ f:α → β, ∀ a:α, P a (f a)) :=
+begin
+  intros A1,
+  let f:α → β := λ a:α, classical.some (A1 a),
+  begin
+    apply exists.intro f,
+    intro a,
+    have A2:P a (classical.some (A1 a)),
+    {
+      apply classical.some_spec,
+    },
+    have A3:f a = (classical.some (A1 a)),
+    {
+      simp,
+    },
+    rw ← A3 at A2,
+    apply A2,
+  end
+end
+
+lemma classical.forall_of_not_exists_not {α:Type*} {P:α → Prop}:(¬(∃ a:α, ¬ P a))→(∀ a:α, P a) :=
+begin
+  intro A1,
+  rw ← @decidable.not_exists_not _ _ (λ x, @classical.prop_decidable _),
+  apply A1,
+end
+
+lemma classical.exists_not_of_not_forall {α:Type*} {P:α → Prop}:(¬(∀ a:α, P a))→ (∃ a:α, ¬ P a) :=
+begin
+  intro A1,
+  apply classical.by_contradiction,
+  intro A2,
+  apply A1,
+  apply classical.forall_of_not_exists_not,
+  apply A2,
+end
+
+
+
+lemma classical.not_forall_iff_exists_not {α:Type*} {P:α → Prop}:(¬(∀ a:α, P a)) ↔(∃ a:α, ¬ P a) :=
+begin
+  split,
+  {
+    apply classical.exists_not_of_not_forall,
+    --apply A1,
+  },
+  {
+    apply not_forall_of_exists_not,
+    --apply A1,
+  },
+end
+
+
+lemma not_exists_iff_forall_not {α:Sort*} {P:α → Prop}:
+    (¬(∃ a:α, P a)) ↔(∀ a:α, ¬ P a) :=
+begin
+  split,
+  {
+    apply forall_not_of_not_exists,
+  },
+  {
+    apply not_exists_of_forall_not,
+    --apply A1,
+  },
+end
+
+
+lemma classical.exists_of_not_forall_not {α:Sort*} {P:α → Prop}:
+    (¬∀ a:α, ¬ P a)  → (∃ a:α, P a) :=
+begin
+  intro A1,
+  cases (classical.em (∃ (a:α), P a)) with A2 A2,
+  {
+    apply A2,
+  },
+  {
+    exfalso,
+    rw not_exists_iff_forall_not at A2,
+    apply A1,
+    apply A2,
+  },
+end
+
+
+lemma canonically_ordered_comm_semiring.add_le_add_right {α:Type*}
+    [canonically_ordered_comm_semiring α] (a b:α):a≤ b → (∀ c, a + c ≤ b + c) :=
+begin
+  intros A1 c,
+  rw add_comm a c,
+  rw add_comm b c,
+  apply canonically_ordered_comm_semiring.add_le_add_left,
+  apply A1,
+end
+
+--This seems to be legitimately new.
+lemma  lt_iff_le_not_eq {α:Type*} [linear_order α] {a b:α}:
+    (a < b) ↔ ((a ≤ b) ∧  (a ≠ b)) :=
+begin
+  split;intros A1,
+  {
+    split,
+    {
+      apply le_of_lt A1,
+    },
+    {
+      apply ne_of_lt A1,
+    },
+  },
+  {
+    rw lt_iff_le_not_le,
+    split,
+    {
+      apply A1.left,
+    },
+    {
+      intro A2,
+      apply A1.right,
+      apply  le_antisymm,
+      apply A1.left,
+      apply A2,
+    },
+  },
+end
+
+lemma le_add_of_nonneg {β:Type*} [ordered_add_comm_monoid β] (a b:β):
+  0 ≤ b → a ≤ a + b :=
+begin
+  intros A1,
+  have B1:a + 0 ≤ a + b,
+  {
+    apply @add_le_add,
+    apply le_refl a,
+    apply A1,
+  },
+  rw add_zero at B1,
+  apply B1,
+end
+
+
+lemma le_add_nonnegative {β:Type*} [canonically_ordered_add_monoid β] (a b:β):
+  a ≤ a + b :=
```

Comment:
This one seems new. How about `le_add_right'`?

Reply:
le_add_right already exists in measure_space.lean according to an error message so I am not able to add le_add_right to ordered_monoid which seems like the best place. How can I avoid getting the error?

Reply:
That's why I said `le_add_right'`

Reply:
```lean
lemma le_add_right' (a b : α) : a ≤ a + b :=
le_add_of_nonneg_right (zero_le _)
```

Reply:
there is already this in `algebra.ordered_monoid`:
```
@[to_additive]
lemma self_le_mul_right (a b : α) : a ≤ a * b :=
le_iff_exists_mul.mpr ⟨b, rfl⟩
```